dagger/dsl/build.py

```diff
@@ -138,6 +135,30 @@ def _parent_serializer_or_default(
     return inputs_from_parent[param_name].serializer
 
 
+def _build_dag_inputs(
+    build_func: Callable,
+    has_a_parent: bool,
+    inputs_from_parent: Mapping[str, NodeInputReference],
+    node_names_by_id: Mapping[str, str],
+):
+    inputs = {}
+    for param_name, param_value in inspect.signature(build_func).parameters.items():
+        if param_name in inputs_from_parent:
+            inputs[param_name] = _build_node_input(
+                inputs_from_parent[param_name],
+                node_names_by_id,
+            )
+        else:
+            inputs[param_name] = FromParam(
+                name="_default_value" if has_a_parent else param_name,
```

Comment:
this is here to avoid setting a param name that exists in the parent, but isn't supposed to reference the parent. For instance:

```python
@dsl.DAG()
def nested_dag(a, b=2):
  f(a, b)

@dsl.DAG()
def dag(a, b):
  nested_dag(a)
```

In this case, we want the nested dag to have a parameter named `b` with a default value of 2, and that default value will always be used in this case.

If we said that `FromParam("b")`, the runtimes would believe it's the same parameter named `b` from the father.

So we need a way to mark that param as "existing, but not really overrideable or linked to anything else". That's why we're prefixing it with `_`, same as the `_hardcoded_value`.

Reply:
makes sense!

Reply:
If that happened, the DSL would generate:

```python
DAG(
    inputs={
        "a": FromParam("a"),
        "b": FromParam("_hardcoded_value", default_value=2),
        "c": FromParam("_hardcoded_value", default_value=20),
        "d": FromParam("_hardcoded_value", default_value=30),
    },
```

The names of the parameters are always unique. They're just referencing a "dummy" parameter from the parent, that should never exist, and therefore it should always go to the default.

The only failure mode I see here is:

* Users explicitly naming a parent parameter `_hardcoded_value`. This is definitely _possible_ in Python, although it would be strange. To prevent this, we could enforce a naming convention for parameters at the "core data structures" level (we're already enforcing a naming convention for nodes).

Reply:
I think it makes sense to enforce some convention for the parameters, e.g. never start with underscore

Reply:
To avoid the failure mode I mentioned previously, I ended up creating a composite name `_hardcoded_value_{param_name}`, to guarantee that if we decided to invoke one of the nested DAGs directly, we could inject each parameter independently